protocol_grpc.go

```diff
@@ -843,28 +843,35 @@ func grpcErrorToTrailer(trailer http.Header, protobuf Codec, err error) {
 	}
 	status := grpcStatusFromError(err)
 	code := strconv.Itoa(int(status.GetCode()))
-	bin, binErr := protobuf.Marshal(status)
-	if binErr != nil {
-		setHeaderCanonical(
-			trailer,
-			grpcHeaderStatus,
-			strconv.FormatInt(int64(CodeInternal), 10 /* base */),
-		)
-		setHeaderCanonical(
-			trailer,
-			grpcHeaderMessage,
-			grpcPercentEncode(
-				fmt.Sprintf("marshal protobuf status: %v", binErr),
-			),
-		)
-		return
+	var bin []byte
+	if len(status.Details) > 0 {
+		// attempt to serialize the Status first, so we can return an error before setting other headers
+		var binErr error
+		bin, binErr = protobuf.Marshal(status)
+		if binErr != nil {
+			setHeaderCanonical(
+				trailer,
+				grpcHeaderStatus,
+				strconv.FormatInt(int64(CodeInternal), 10 /* base */),
+			)
+			setHeaderCanonical(
+				trailer,
+				grpcHeaderMessage,
+				grpcPercentEncode(
+					fmt.Sprintf("marshal protobuf status: %v", binErr),
+				),
+			)
+			return
+		}
 	}
 	if connectErr, ok := asError(err); ok {
 		mergeHeaders(trailer, connectErr.meta)
 	}
```

Comment:
I wonder if we should move this up, so we can always propagate the user's trailer metadata into the response, even if we fail to marshal the status proto.

Reply:
Could potentially simplify to:
```go
	var (
		status  = grpcStatusFromError(err)
		code    = status.GetCode()
		message = status.GetMessage()
		bin     []byte
	)
	if connectErr, ok := asError(err); ok && !connectErr.wireErr {
		mergeMetadataHeaders(trailer, connectErr.meta)
	}
	if len(status.Details) > 0 {
		var binErr error
		bin, binErr = protobuf.Marshal(status)
		if binErr != nil {
			code = int32(CodeInternal)
			message = fmt.Sprintf("marshal protobuf status: %v", binErr)
		}
	}
```

Reply:
Happy to do that. I'm not super familiar with how folks use the header metadata in errors, but it's possible that you *wouldn't* want them to be emitted in the case of a marshalling failure (e.g. if they were cache control headers or something).